app/controllers/formbuilder/edit.js

```diff
@@ -108,36 +84,77 @@ export default class FormbuilderEditController extends Controller {
     );
 
     this.builderStore.registerObserver(() => {
-      this.serializeSourceToTtl();
+      this.handleBuilderFormChange.perform();
     });
+  });
 
-    if (isInitialRouteCall) {
-      this.setFormChanged(false);
-      this.isAddingValidationToForm = false;
-      this.isShowBuilder = true;
-    }
+  setupPreviewForm = restartableTask(async () => {
+    // force a component recreation by unrendering it very briefly
+    // Ideally the RdfForm component would do the right thing when the formStore
+    // and form arguments change, but we're not there yet.
+    await timeout(1);
 
-    this.isInitialDataLoaded = true;
-  }
+    this.previewStore = new ForkingStore();
+    this.previewStore.parse(
+      this.formCode,
+      GRAPHS.formGraph.value,
+      'text/turtle'
+    );
+
+    this.previewForm = this.previewStore.any(
+      undefined,
+      RDF('type'),
+      FORM('Form'),
+      GRAPHS.formGraph
+    );
+  });
 
   @action
   setFormChanged(value) {
     this.formChanged = value;
   }
 
   @action
-  serializeSourceToTtl() {
+  async handleCodeChange(newCode) {
+    this.formChanged = true;
+    this.formCode = newCode;
+    this.setupForms();
+  }
+
+  handleBuilderFormChange = restartableTask(async () => {
+    await timeout(1); // small timeout to group multiple store observer callbacks together
```

Comment:
I am not getting this comment.. If I read it correctly: this is a restartable task and you want to wait 1ms inbetween restart to make sure the observables are ready?